Migration failure on Oracle Standard Edition During CREATE INDEX Step

Problem Statement
Migration 2022.05.R01 to 2022.08.R01 fails on Oracle Standard Edition during CREATE INDEX step

Steps to Reproduce
Install 2022.05.R01 using Oracle Standard Edition
Upgrade to 2022.08.R01
Observed Results
```
Caused by: org.springframework.jdbc.UncategorizedSQLException: PreparedStatementCallback; uncategorized SQLException for SQL [create index IDX_EMPI_MATCH_TGT_VER on MPI_LINK(MATCH_RESULT, TARGET_PID, VERSION) ONLINE DEFERRED INVALIDATION]; SQL state [67000]; error code [439]; ORA-00439: feature not enabled: Online Index Build
; nested exception is java.sql.SQLException: ORA-00439: feature not enabled: Online Index Build
```

Expected Results
No errors

Environment (please complete the following information):
hapi-fhir 6.1.0 upgrade
Oracle DB Server Standard Edition

Oracle DB Server Standard Edition doesn't support ONLINE option in the CREATE INDEX command

--- a/hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java
+++ b/hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java
@@ -126,22 +126,24 @@ public class AddIndexTask extends BaseTableTask {
 			mssqlWhereClause = buildMSSqlNotNullWhereClause();
 		}
 		// Should we do this non-transactionally?  Avoids a write-lock, but introduces weird failure modes.
-		String postgresOnline = "";
-		String oracleOnlineDeferred = "";
+		String postgresOnlineClause = "";
+		String msSqlOracleOnlineClause = "";
 		if (myOnline) {
 			switch (getDriverType()) {
 				case POSTGRES_9_4:
 				case COCKROACHDB_21_1:
-					postgresOnline = "CONCURRENTLY ";
+					postgresOnlineClause = "CONCURRENTLY ";
 					// This runs without a lock, and can't be done transactionally.
 					setTransactional(false);
 					break;
 				case ORACLE_12C:
-					oracleOnlineDeferred = " ONLINE DEFERRED INVALIDATION";
+					if (myMetadataSource.isOnlineIndexSupported(getConnectionProperties())) {
+						msSqlOracleOnlineClause = " ONLINE DEFERRED INVALIDATION";
+					}
 					break;
 				case MSSQL_2012:
 					if (myMetadataSource.isOnlineIndexSupported(getConnectionProperties())) {
-						oracleOnlineDeferred = " WITH (ONLINE = ON)";
+						msSqlOracleOnlineClause = " WITH (ONLINE = ON)";
 					}
 					break;
 				default:
@@ -150,8 +152,8 @@ public class AddIndexTask extends BaseTableTask {
 
 
 		String sql =
-			"create " + unique + "index " + postgresOnline + myIndexName +
-			" on " + getTableName() + "(" + columns + ")" + includeClause +  mssqlWhereClause + oracleOnlineDeferred;
+			"create " + unique + "index " + postgresOnlineClause + myIndexName +
+			" on " + getTableName() + "(" + columns + ")" + includeClause +  mssqlWhereClause + msSqlOracleOnlineClause;
 		return sql;
 	}
 

--- a/hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java
+++ b/hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java
@@ -34,20 +34,22 @@ public class MetadataSource {
 	public boolean isOnlineIndexSupported(DriverTypeEnum.ConnectionProperties theConnectionProperties) {
 
 		switch (theConnectionProperties.getDriverType()) {
-			case ORACLE_12C:
 			case POSTGRES_9_4:
 			case COCKROACHDB_21_1:
 				return true;
 			case MSSQL_2012:
-				String edition = getEdition(theConnectionProperties);
-				return edition.startsWith("Enterprise");
+				String mssqlEdition = getEdition(theConnectionProperties);
+				return mssqlEdition.startsWith("Enterprise");
+			case ORACLE_12C:
+				String oracleEdition = getEdition(theConnectionProperties);
+				return oracleEdition.contains("Enterprise");
 			default:
 				return false;
 		}
 	}
 
 	/**
-	 * Get the MS Sql Server edition.  Other databases are not supported yet.
+	 * Get the MS Sql Server or Oracle Server edition.  Other databases are not supported yet.
 	 *
 	 * @param theConnectionProperties the database to inspect
 	 * @return the edition string (e.g. Standard, Enterprise, Developer, etc.)
@@ -56,6 +58,8 @@ public class MetadataSource {
 		final String result;
 		if (theConnectionProperties.getDriverType() == DriverTypeEnum.MSSQL_2012) {
 			result = theConnectionProperties.newJdbcTemplate().queryForObject("SELECT SERVERPROPERTY ('edition')", String.class);
+		} else if (theConnectionProperties.getDriverType() == DriverTypeEnum.ORACLE_12C) {
+			result = theConnectionProperties.newJdbcTemplate().queryForObject("SELECT * FROM v$version WHERE banner LIKE 'Oracle%';", String.class);
 		} else {
 			throw new UnsupportedOperationException(Msg.code(2084) + "We only know about MSSQL editions.");
 		}

--- a/hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTaskTest.java
+++ b/hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTaskTest.java
@@ -217,6 +217,22 @@ public class AddIndexTaskTest extends BaseTest {
 					assertEquals("create index IDX_ANINDEX on SOMETABLE(PID, TEXTCOL)", mySql);
 				}
 			}
+
+			@ParameterizedTest(name = "{index}: {0}")
+			@ValueSource(booleans = { true, false } )
+			public void offForUnsupportedVersionsOfOracleServer(boolean theSupportedFlag) {
+				myTask.setDriverType(DriverTypeEnum.ORACLE_12C);
+				myTask.setOnline(true);
+				myTask.setMetadataSource(mockMetadataSource);
+				Mockito.when(mockMetadataSource.isOnlineIndexSupported(Mockito.any())).thenReturn(theSupportedFlag);
+
+				mySql = myTask.generateSql();
+				if (theSupportedFlag) {
+					assertEquals("create index IDX_ANINDEX on SOMETABLE(PID, TEXTCOL) ONLINE DEFERRED INVALIDATION", mySql);
+				} else {
+					assertEquals("create index IDX_ANINDEX on SOMETABLE(PID, TEXTCOL)", mySql);
+				}
+			}
 		}
 	}
 }

